internal/queue/consumer.go

```diff
@@ -0,0 +1,288 @@
+	"strconv"
+	"time"
+
+	"github.com/canonical/github-runner-operators/internal/database"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+const platform = "github"
+
+// JobDatabase is a small interface that matches the relevant method on internal/database.Database.
+type JobDatabase interface {
+	AddJob(ctx context.Context, job *database.Job) error
+	UpdateJobStarted(ctx context.Context, platform, id string, startedAt time.Time, raw map[string]any) error
+	UpdateJobCompleted(ctx context.Context, platform, id string, completedAt time.Time, raw map[string]any) error
+}
+
+// AmqpConsumer is an AMQP consumer for workflow job events.
+type AmqpConsumer struct {
+	client    *Client
+	queueName string
+	db        JobDatabase
+	logger    *slog.Logger
+}
+
+// NewAmqpConsumer creates a new AmqpConsumer with the given dependencies.
+func NewAmqpConsumer(uri, queueName string, db JobDatabase, logger *slog.Logger) *AmqpConsumer {
+	if logger == nil {
+		logger = slog.Default()
+	}
+	return &AmqpConsumer{
+		queueName: queueName,
+		db:        db,
+		logger:    logger,
+		client: &Client{
+			uri:         uri,
+			connectFunc: amqpConnect,
+		},
+	}
+}
+
+// Start begins consuming messages from the AMQP queue and processing them.
+func (c *AmqpConsumer) Start(ctx context.Context) error {
+	err := c.resetConnectionOrChannelIfNecessary(ctx)
+	if err != nil {
+		return err
+	}
+
+	msgs, err := c.consumeMsgs()
+	if err != nil {
+		return err
+	}
+
+	c.logger.Info("AMQP consumer started", "queue", c.queueName)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case msg, ok := <-msgs:
+			if !ok {
+				c.logger.Warn("messages channel closed, retrying connection")
+				if err := c.resetConnectionOrChannelIfNecessary(ctx); err != nil {
+					return err
+				}
+				msgs, err = c.consumeMsgs()
+				if err != nil {
+					return err
+				}
+				continue
+			}
+
+			err := c.handleMessage(ctx, msg.Body)
+			if err == nil {
+				msg.Ack(false)
+				continue
+			}
+
+			var errHandle *MessageHandlingError
+			if errors.As(err, &errHandle) && !errHandle.Requeue {
+				msg.Nack(false, false) // don't requeue
+				c.logger.Error("failed to handle message without requeue", "error", err)
+				continue
+			}
+			// Other errors are requeued by default
+			c.logger.Error("failed to handle message, requeuing", "error", err)
+			msg.Nack(false, true) // requeue
+		}
+	}
+}
+
+// resetConnectionOrChannelIfNecessary resets the AMQP connection or channel if they are nil or closed.
+// It includes a retry mechanism with exponential backoff.
+func (c *AmqpConsumer) resetConnectionOrChannelIfNecessary(ctx context.Context) error {
+	backoff := time.Second
+	maxBackoff := time.Minute
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		c.client.mu.Lock() // Lock to prevent concurrent access to connection/channel object
+
+		var err error
+		if err = c.client.resetConnection(); err != nil {
+			c.client.mu.Unlock()
+			goto retry
```

Comment:
Using "goto" statements is generally discouraged, as it leads to hard to follow control flow.